src/Repository/Pdo/ClientRepository.php

```diff
@@ -20,39 +21,69 @@ class ClientRepository extends AbstractRepository implements ClientRepositoryInt
     /**
      * {@inheritDoc}
      */
-    public function getClientEntity(
-        $clientIdentifier,
-        $grantType = null,
-        $clientSecret = null,
-        $mustValidateSecret = true
-    ) {
-        $sth = $this->pdo->prepare(
-            'SELECT * FROM oauth_clients WHERE name = :clientIdentifier'
+    public function getClientEntity($clientIdentifier) : ?ClientEntityInterface
+    {
+        $clientData = $this->getClientData($clientIdentifier);
+
+        if (empty($clientData)) {
+            return null;
+        }
+
+        return new ClientEntity(
+            $clientIdentifier,
+            $clientData['name'] ?? '',
+            $clientData['redirect'] ?? ''
         );
-        $sth->bindParam(':clientIdentifier', $clientIdentifier);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function validateClient($clientIdentifier, $clientSecret, $grantType) : bool
+    {
+        $clientData = $this->getClientData($clientIdentifier);
+
+        if (empty($clientData)) {
+            return false;
+        }
+
+        if (! $this->isGranted($clientData, $grantType)) {
+            return false;
+        }
 
-        if (false === $sth->execute()) {
-            return;
+        if (empty($clientData['secret']) || ! password_verify((string) $clientSecret, $clientData['secret'])) {
+            return false;
         }
-        $row = $sth->fetch();
-        if (empty($row) || ! $this->isGranted($row, $grantType)) {
-            return;
+
+        return true;
+    }
+
+    protected function getClientData(string $clientIdentifier) : ?array
```

Comment:
Is there a reason why this is protected an not private?

Reply:
Copy-paste mistake. I'll fix it.